header/p2p/exchange.go

```diff
@@ -288,3 +259,57 @@ func convertStatusCodeToError(code p2p_pb.StatusCode) error {
 		return fmt.Errorf("unknown status code %d", code)
 	}
 }
+
+// sendMessage opens the stream to the given peers and sends ExtendedHeaderRequest to fetch ExtendedHeaders.
+func sendMessage(
+	ctx context.Context,
+	host host.Host,
+	to peer.ID,
+	protocol protocol.ID,
+	req *p2p_pb.ExtendedHeaderRequest,
+) ([]*p2p_pb.ExtendedHeaderResponse, uint64, uint64, error) {
+	startTime := time.Now()
+	stream, err := host.NewStream(ctx, to, protocol)
+	if err != nil {
+		return nil, 0, 0, err
+	}
+
+	// set stream deadline from the context deadline.
+	// if it is empty, then we assume that it will
+	// hang until the server will close the stream by the timeout.
+	if dl, ok := ctx.Deadline(); ok {
+		if err = stream.SetDeadline(dl); err != nil {
+			log.Debugw("error setting deadline: %s", err)
+		}
+	}
+	// send request
+	_, err = serde.Write(stream, req)
+	if err != nil {
+		stream.Reset() //nolint:errcheck
+		return nil, 0, 0, err
+	}
+	err = stream.CloseWrite()
+	if err != nil {
+		log.Error(err)
```

Comment:
We cannot ignore this error, iirc. If it errors, the stream is unusable

Reply:
But we do not need stream anymore. Error can appear on closing the stream

Reply:
No, we need to read the data out of it

Reply:
So, if we are getting error here, then we should immediately return it?

Reply:
Yes